An implementation of the isogram exercise

AtelesPaniscus opened the pull request:
Enjoy.

marnen commented on exercises/isogram/isogram-example.apl:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env apl --script
+
+version ← '1.2.0' ⍝ isogram exercise, GNU-APL track, exercism
```
What is this for?

marnen commented on exercises/isogram/isogram.tc:
```diff
@@ -0,0 +1,49 @@
+  1⎕CR	(version)
+'1.2.0'
+
+⍝ ∇isIsogram
```
I've been using underscore_case for the rest of the code, which appears to be a weak preference in modern APL (makes sense, since APL didn't originally support more than one case of letters).

marnen commented on exercises/isogram/isogram.tc:
```diff
@@ -0,0 +1,49 @@
+
+⍝ ∇isIsogram
+
+⍝ empty string
```
I assume these are all the test cases from the canonical JSON file?

marnen commented on exercises/isogram/isogram.tc:
```diff
@@ -0,0 +1,49 @@
+⍝ ∇isIsogram
+
+⍝ empty string
+  1⎕CR	(isIsogram '')
```
The parentheses are neither necessary nor idiomatic for APL. Functions in APL have "long right scope", which means that everything after a function's name is considered as its right argument.

marnen commented on exercises/isogram/isogram-example.apl:
```diff
@@ -0,0 +1,10 @@
+alphabet ← 'abcdefghijklmnopqrstuvwxyz'
+ALPHABET ← 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+
+counts ← {+/ ⍺ ∘.= ⍵}
```
As much as I like the idea of APL lambdas, I'd prefer that we not use them in Exercism sample code, since they're not part of the APL standard, and I would like to keep the sample code as close to the standard as possible to facilitate use of other APL interpreters than GNU APL.

In this case, it's easy enough to use `∇` syntax to define the function:

```apl
∇result←alphabet counts string
  +/ alphabet ∘.= string
∇
```

marnen commented on exercises/isogram/isogram-example.apl:
```diff
@@ -0,0 +1,10 @@
+
+counts ← {+/ ⍺ ∘.= ⍵}
+
+isIsogram ← {2 > ⌈/ (alphabet counts ⍵) + (ALPHABET counts ⍵)}
```
I'm not at all sure that I like this implementation; I think I'd prefer to normalize case and count. But I'll think about it further.

marnen commented:
Thanks very much for this pull request! I've left a few comments and suggestions, but I really appreciate the contribution.

marnen commented:
@AtelesPaniscus Do you no longer want to get this merged?

AtelesPaniscus commented:
Is it really two weeks since I deleted my branch ?  My apologies for not replying sooner.

I deleted the branch because at the time I was poised to update it and I reckoned that would update the pull request and I didn't want you wasting your time writing any more 'comments and suggestions'.   It really didn't occur to me that you would be notified.  Silly of me.  My apologises.

You don't seem to have the time and inclination to contribute exercises to the track but you do have both when it comes to 'comments and suggestions' of other people's contributions.

I don't have the time because I really don't have the time any more.  As for inclination.  This is something I do as R&R.  If I don't enjoy it, well,  it is not that I refuse to do it.  It is just that something else will find me and I'll enjoy doing that instead.

I appreciate Open Source and the idea of giving something back/  Here was something I had done.  If you want it «tant mieux»; if you don't «tant pis».  What puzzles me with those with high standards is why they don't take and change what they don't like to something that suits.  After all, once it is Open Source, anyone can take it and change it to something I would find shameful and all they have to do is give me credit.  That is the nature of the beast.

The older generation here has a saying "If you want something done properly, do it yourself."  I don't like the word 'properly'.  It implies other people are incompetent just because they are not you.  In my opinion if you want something just the way you like it you either go a very expensive restaurant and get yourself a reputation or you cook it yourself.

I wrote (am still writing) a code generator that produces test cases from canonical data.  You saw an early example.  Each time I hit a new snag I have to consider how to adapt the generator.  It would have been nice to consult.  I expect the choices I have made on my own mean the code generated is less and less to your taste.

To test the code generator I write example solutions. There are now some 31.  The Exercism guidelines state these do not have to be great code.  Seems I got the wrong end of the stick.

I have published the exercises on GitHub.  There are even notes on the issues that have determined the direction the generator has taken but I do not have explain to you or anyone else why the example solutions are the way they are.

You take care now.

marnen commented:
@AtelesPaniscus:

> I deleted the branch because at the time I was poised to update it and I reckoned that would update the pull request and I didn't want you wasting your time writing any more 'comments and suggestions'. It really didn't occur to me that you would be notified. Silly of me. My apologises.

Understood. In the future, you may want to be aware of these two things:

* GitHub will automatically update your PR if you push more commits to the branch that it was made from; no need to close.
* If you do close your own pull request, a comment explaining why is helpful in order to avoid confusion.

> You don't seem to have the time and inclination to contribute exercises to the track but you do have both when it comes to 'comments and suggestions' of other people's contributions.

This is a hobby project for me, and I don't have a lot of time to work on it, true. At the same time, though, I would consider it rude to contributors to leave a pull request without a response for long, so I try hard to respond to pull requests as soon as I receive them. That's just common courtesy as far as I'm concerned.

> I appreciate Open Source and the idea of giving something back/ Here was something I had done. If you want it «tant mieux»; if you don't «tant pis». What puzzles me with those with high standards is why they don't take and change what they don't like to something that suits. After all, once it is Open Source, anyone can take it and change it to something I would find shameful and all they have to do is give me credit. That is the nature of the beast.

Because it's open source, you are welcome to maintain your own fork of the code that you control, and you are welcome to put anything you like in that fork. But a pull request is something different.

A pull request is a request for a maintainer to incorporate code that you've written into a repository that that maintainer controls. As such, the maintainer has the responsibility to make sure that the code is ready for inclusion before merging it in and taking responsibility for it.

> I wrote (am still writing) a code generator that produces test cases from canonical data. 

Awesome! Where is the repository for that project?

> You saw an early example. Each time I hit a new snag I have to consider how to adapt the generator. 

Or you can ask for help. That's another advantage of open source. (Heck, if you open an issue and tag me, I'll try to help as time permits.)

> It would have been nice to consult. 

Yes, exactly! That's why I left review comments. If you wanted my input on the generator, you could have asked me at any point in the process. You still could ask me, in fact; I'd be happy to provide such input.

> I expect the choices I have made on my own mean the code generated is less and less to your taste.

That would not be a reasonable assumption as stated.

> To test the code generator I write example solutions. There are now some 31. The Exercism guidelines state these do not have to be great code. Seems I got the wrong end of the stick.

The Exercism guidelines do state that they don't have to be great code. But I would at least like them to run on as many APL interpreters as possible, which means avoiding nonstandard features like lambdas.

If you submitted an exercise containing lambdas as a "slugger"/student/whatever one calls it, I would have no problem with it. I'm just really trying to avoid interpreter-specific code in the track repo itself, except where absolutely necessary.

> I have published the exercises on GitHub. 

Where?

> There are even notes on the issues that have determined the direction the generator has taken but I do not have explain to you or anyone else why the example solutions are the way they are.

You don't have to explain that code to anyone else if you're just planning on keeping it in your fork, true. But as stated above, if you want a maintainer to incorporate code you've written into his repo, you should be prepared to discuss why that code is the way it is: the maintainer needs to understand the code and be prepared to take responsibility for it going forward.

I am the maintainer for this particular repo. I asked some reasonable questions about your implementation prior to merging it in and taking responsibility for it. Any maintainer (that I've dealt with) will do likewise.

> You take care now.

That came out condescending. Was it meant that way? If so, please find a different way of interacting with me in the future. (If it wasn't meant to be condescending, be aware that it came out that way, and please consider how to avoid false impressions of your intent in the future.)
